tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
+            'last class in attribute is to be removed' => [
+                'html' => '<p class="foo bar baz">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+                'classes expected to be removed' => ['baz'],
+            ],
+        ];
+        $matchedClassesWithExtraWhitespaceData = $this->matchedClassesWithExtraWhitespaceDataProvider();
+
+        return $nonMatchedClassesData + $matchedClassesData + $matchedClassesWithExtraWhitespaceData;
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function matchedClassesWithExtraWhitespaceDataProvider()
+    {
+        return [
+            '1 attribute, 1 class with extra whitespace, that class to be kept' => [
+                'html' => '<p class=" foo ">hello</p>',
+                'classes to keep' => ['foo'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, both to be kept' => [
+                'html' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, 1 to be kept' => [
+                'html' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, both to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, 1 to be kept' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['foo'],
+                'classes expected to be removed' => ['bar'],
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string[] $classesToKeep
+     * @param string[] $classesExpectedToBeRemoved
+     *
+     * @dataProvider classesDataProvider
+     */
+    public function removeRedundantClassesRemovesOnlyClassesNotToKeep(
+        $html,
+        array $classesToKeep,
+        array $classesExpectedToBeRemoved = []
+    ) {
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        $result = $subject->render();
+        foreach ($classesToKeep as $class) {
+            $expectedInstanceCount = \substr_count($html, $class);
+            if ($expectedInstanceCount !== 0) {
+                self::assertSame(
+                    $expectedInstanceCount,
+                    \substr_count($result, $class),
+                    'asserting \'' . $result . '\' contains ' . $expectedInstanceCount . ' instance(s) of "' . $class
+                    . '"'
+                );
+            }
+        }
+        foreach ($classesExpectedToBeRemoved as $class) {
+            self::assertNotContains($class, $result);
+        }
```

Comment:
Please let's split this up into separate tests.

Reply:
OK.  We can have `removeRedundantClassesRemovesClassesNotToKeep` and `removeRedundantClassesNotRemovesClassesNotToKeep`.  It may then be realized that not all the datasets are applicable to one or other, which might help with

> Instead of having a data provider with a different kinds of data sets, I propose we have three separate tests that use each data subset and focus on one aspect of the behavior.